// typescript-formatter/lib/index.js
"use strict";
var formatter_1 = require("./formatter");
var utils_1 = require("./utils");
var fs = require("fs");
var base_1 = require("./provider/base");
var tsconfigjson_1 = require("./provider/tsconfigjson");
var editorconfig_1 = require("./provider/editorconfig");
var tslintjson_1 = require("./provider/tslintjson");
function processFiles(files, opts) {
    "use strict";
    var resultMap = {};
    var promises = files.map(function (fileName) {
        if (!fs.existsSync(fileName)) {
            var result = {
                fileName: fileName,
                options: null,
                message: fileName + " does not exist. process abort.\n",
                error: true,
                src: "",
                dest: "",
            };
            return Promise.resolve(result);
        }
        var content = fs.readFileSync(fileName).toString();
        return processString(fileName, content, opts);
    });
    return Promise.all(promises).then(function (resultList) {
        resultList.forEach(function (result) {
            resultMap[result.fileName] = result;
        });
        return resultMap;
    });
}
exports.processFiles = processFiles;
function processStream(fileName, input, opts) {
    "use strict";
    input.setEncoding("utf8");
    var promise = new Promise(function (resolve, reject) {
        var fragment = "";
        input.on("data", function (chunk) {
            fragment += chunk;
        });
        input.on("end", function () {
            resolve(fragment);
        });
    });
    return promise.then(function (content) { return processString(fileName, content, opts); });
}
exports.processStream = processStream;
function processString(fileName, content, opts) {
    "use strict";
    var formatOptions = utils_1.createDefaultFormatCodeOptions();
    var optGenPromises = [];
    var postProcesses = [];
    if (opts.tsfmt) {
        optGenPromises.push(base_1.default(fileName, opts, formatOptions));
    }
    if (opts.tsconfig) {
        optGenPromises.push(tsconfigjson_1.default(fileName, opts, formatOptions));
    }
    if (opts.editorconfig) {
        optGenPromises.push(editorconfig_1.default(fileName, opts, formatOptions));
    }
    if (opts.tslint) {
        optGenPromises.push(tslintjson_1.default(fileName, opts, formatOptions));
        postProcesses.push(tslintjson_1.postProcess);
    }
    return Promise
        .all(optGenPromises)
        .then(function () {
        var formattedCode = formatter_1.default(fileName, content, formatOptions);
        if (formattedCode.trimRight) {
            formattedCode = formattedCode.trimRight();
            formattedCode += formatOptions.NewLineCharacter;
        }
        postProcesses.forEach(function (postProcess) {
            formattedCode = postProcess(fileName, formattedCode, opts, formatOptions) || formattedCode;
        });
        formattedCode = formattedCode.replace(/\r?\n/g, formatOptions.NewLineCharacter);
        var message;
        var error = false;
        if (opts && opts.verify) {
            if (content !== formattedCode) {
                message = fileName + " is not formatted\n";
                error = true;
            }
        }
        else if (opts && opts.replace) {
            if (content !== formattedCode) {
                fs.writeFileSync(fileName, formattedCode);
                message = "replaced " + fileName + "\n";
            }
        }
        else if (opts && !opts.dryRun) {
            message = formattedCode;
        }
        var result = {
            fileName: fileName,
            options: formatOptions,
            message: message,
            error: error,
            src: content,
            dest: formattedCode,
        };
        return Promise.resolve(result);
    });
}
exports.processString = processString;
