// raptor-logging/lib/raptor-logging.js
var EMPTY_FUNC = function() {
        return false;
    },
    /**
     * @name raptor/logging/voidLogger
     */
    voidLogger = {

        /**
         *
         */
        isTraceEnabled: EMPTY_FUNC,

        /**
         *
         */
        isDebugEnabled: EMPTY_FUNC,

        /**
         *
         */
        isInfoEnabled: EMPTY_FUNC,

        /**
         *
         */
        isWarnEnabled: EMPTY_FUNC,

        /**
         *
         */
        isErrorEnabled: EMPTY_FUNC,

        /**
         *
         */
        isFatalEnabled: EMPTY_FUNC,

        /**
         *
         */
        dump: EMPTY_FUNC,

        /**
         *
         */
        trace: EMPTY_FUNC,

        /**
         *
         */
        debug: EMPTY_FUNC,

        /**
         *
         */
        info: EMPTY_FUNC,

        /**
         *
         */
        warn: EMPTY_FUNC,

        /**
         *
         */
        error: EMPTY_FUNC,

        /**
         *
         */
        fatal: EMPTY_FUNC
    };

var stubs = {
    /**
     *
     * @param className
     * @returns
     */
    logger: function() {
        return voidLogger;
    },

    configure: EMPTY_FUNC,

    voidLogger: voidLogger
};


module.exports = stubs;

// Trick the JavaScript module bundler so that it doesn't include the implementation automatically
var RAPTOR_LOGGING_IMPL = './raptor-logging-impl';

if (!process.browser) {
    var implPath;

    try {
        implPath = require.resolve(RAPTOR_LOGGING_IMPL);
    } catch(e) {
        /*
        Fixes https://github.com/raptorjs/raptor-logging/issues/4
        If `./raptor-logging-impl` is unable to be loaded then it means that a server bundle was built and it does
        not support dynamic requires since the server bundle is being loaded from a different
        directory that breaks the relative path.
        */
    }
    if (implPath) {
        require(implPath);
    }
}
