// raptor-dom/raptor-dom-client.js
var raptorPubsub = require('raptor-pubsub');

function getNode(el) {
    if (typeof el === 'string') {
        var elId = el;
        el = document.getElementById(elId);
        if (!el) {
            throw new Error('Target element not found: "' + elId + '"');
        }
    }
    return el;
}

function _beforeRemove(referenceEl) {
    if (raptorPubsub) {
        raptorPubsub.emit('dom/beforeRemove', {
            el: referenceEl
        });
    }
}

var dom = {
    forEachChildEl: function(node, callback, scope) {
        dom.forEachChild(node, callback, scope, 1);
    },
    forEachChild: function(node, callback, scope, nodeType) {
        if (!node) {
            return;
        }
        var i = 0;
        var childNodes = node.childNodes;
        var len = childNodes.length;
        for (; i < len; i++) {
            var childNode = childNodes[i];
            if (childNode && (nodeType == null || nodeType == childNode.nodeType)) {
                callback.call(scope, childNode);
            }
        }
    },
    detach: function(child) {
        child = getNode(child);
        child.parentNode.removeChild(child);
    },
    appendTo: function(newChild, referenceParentEl) {
        getNode(referenceParentEl).appendChild(getNode(newChild));
    },
    remove: function(el) {
        el = getNode(el);
        _beforeRemove(el);
        if (el.parentNode) {
            el.parentNode.removeChild(el);
        }
    },
    removeChildren: function(parentEl) {
        parentEl = getNode(parentEl);

        var i = 0;
        var childNodes = parentEl.childNodes;
        var len = childNodes.length;
        for (; i < len; i++) {
            var childNode = childNodes[i];
            if (childNode && childNode.nodeType === 1) {
                _beforeRemove(childNode);
            }
        }
        parentEl.innerHTML = '';
    },
    replace: function(newChild, replacedChild) {
        replacedChild = getNode(replacedChild);
        _beforeRemove(replacedChild);
        replacedChild.parentNode.replaceChild(getNode(newChild), replacedChild);
    },
    replaceChildrenOf: function(newChild, referenceParentEl) {
        referenceParentEl = getNode(referenceParentEl);
        dom.forEachChildEl(referenceParentEl, function(childEl) {
            _beforeRemove(childEl);
        });
        referenceParentEl.innerHTML = '';
        referenceParentEl.appendChild(getNode(newChild));
    },
    insertBefore: function(newChild, referenceChild) {
        referenceChild = getNode(referenceChild);
        referenceChild.parentNode.insertBefore(getNode(newChild), referenceChild);
    },
    insertAfter: function(newChild, referenceChild) {
        referenceChild = getNode(referenceChild);
        newChild = getNode(newChild);
        var nextSibling = referenceChild.nextSibling;
        var parentNode = referenceChild.parentNode;
        if (nextSibling) {
            parentNode.insertBefore(newChild, nextSibling);
        } else {
            parentNode.appendChild(newChild);
        }
    },
    prependTo: function(newChild, referenceParentEl) {
        referenceParentEl = getNode(referenceParentEl);
        referenceParentEl.insertBefore(getNode(newChild), referenceParentEl.firstChild || null);
    }
};

/*
var jquery = window.$;
if (!jquery) {
    try {
        jquery = require('jquery');
    }
    catch(e) {}
}

if (jquery) {
    dom.ready = function(callback, thisObj) {
        jquery(function() {
            callback.call(thisObj);
        });
    };
} else {
    dom.ready = require('./raptor-dom_documentReady');
}
*/
dom.ready = require('./ready');

module.exports = dom;
